pkg/ingester/owned_series.go

```diff
@@ -0,0 +1,183 @@
+		getIngesterShardSize: getIngesterShardSize,
+		getTSDBUsers:         getTSDBUsers,
+		getTSDB:              getTSDB,
+	}
+
+	oss.Service = services.NewTimerService(interval, oss.starting, oss.iter, nil)
+	return oss
+}
+
+// This is Starting function for ownedSeries service. Service is only started after all TSDBs are opened.
+// Pushes are not allowed yet when this function runs.
+func (oss *ownedSeriesService) starting(ctx context.Context) error {
+	err := ring.WaitInstanceState(ctx, oss.ingestersRing, oss.instanceID, ring.ACTIVE)
+	if err != nil {
+		return err
+	}
+
+	if _, err := oss.checkRingForChanges(); err != nil {
+		return fmt.Errorf("can't read ring: %v", err)
+	}
+
+	// We pass ringChanged=true, but all TSDBs at this point (after opening TSDBs, but before ingester switched to Running state) also have "new user" trigger set anyway.
+	oss.updateAll(ctx, true)
+	return nil
+}
+
+// This function runs periodically. It checks if ring has changed, and updates number of owned series for any
+// user that requires it (due to ring change, compaction, shard size change, ...).
+func (oss *ownedSeriesService) iter(ctx context.Context) error {
+	ringChanged, err := oss.checkRingForChanges()
+	if err != nil {
+		level.Error(oss.logger).Log("msg", "can't check ring for updates", "err", err)
+		return nil // If we returned error, service would stop.
+	}
+
+	start := time.Now()
+	updatedUsers := oss.updateAll(ctx, ringChanged)
+	if updatedUsers > 0 {
+		level.Info(oss.logger).Log("msg", "updated owned series for users", "updatedUsers", updatedUsers, "duration", time.Since(start), "ringChanged", ringChanged)
+	}
+	return nil
+}
+
+// Reads current ring, stores it, and returns bool indicating whether ring has changed since last call of this method.
+func (oss *ownedSeriesService) checkRingForChanges() (bool, error) {
+	rs, err := oss.ingestersRing.GetAllHealthy(ownedSeriesRingOp)
+	if err != nil {
+		return false, err
+	}
+
+	// Since token ranges computation doesn't care about state, we don't need to either.
+	ringChanged := ring.HasReplicationSetChangedWithoutState(oss.previousRing, rs)
+	oss.previousRing = rs
+	return ringChanged, nil
+}
+
+// updateAll iterates over all open TSDBs and updates owned series for all users that need it, either
+// because of external trigger (new user, compaction), or because of changed token ranges.
+func (oss *ownedSeriesService) updateAll(ctx context.Context, ringChanged bool) int {
+	updatedUsers := 0
+	for _, userID := range oss.getTSDBUsers() {
+		if ctx.Err() != nil {
+			return updatedUsers
+		}
+
+		db := oss.getTSDB(userID)
+		if db == nil {
+			continue
+		}
+
+		if oss.updateTenant(userID, db, ringChanged) {
+			updatedUsers++
+		}
+	}
+	return updatedUsers
+}
+
+// Updates token ranges and recomputes owned series for user, if necessary. If recomputation happened, true is returned.
+func (oss *ownedSeriesService) updateTenant(userID string, db *userTSDB, ringChanged bool) bool {
```

Comment:
I agree it is cumbersome because it covers many different scenarios. I've extended the comment to explain which scenarios we cover, to help understand why it is like this.